src/robot/libdocpkg/model.py

```diff
@@ -190,3 +228,155 @@ def _arg_to_dict(self, arg):
             'required': arg.required,
             'repr': unicode(arg)
         }
+
+
+class DataTypeCatalog(object):
+
+    def __init__(self):
+        self._enum_catalog = set()
+        self._typed_dict_catalog = set()
+
+    def __iter__(self):
+        for type_dooc in sorted(self._typed_dict_catalog | self._enum_catalog):
+            yield type_dooc
+
+    def __bool__(self):
+        return bool(self._enum_catalog or self._typed_dict_catalog)
+
+    __nonzero__ = __bool__
+
+    @property
+    def enum_catalog(self):
+        return sorted(self._enum_catalog)
+
+    @property
+    def typed_dict_catalog(self):
+        return sorted(self._typed_dict_catalog)
+
+    def add(self, typ):
+        type_doc = self._get_type_doc_object(typ)
+        if isinstance(type_doc, EnumDoc):
+            self._enum_catalog.add(type_doc)
+        elif isinstance(type_doc, TypedDictDoc):
+            self._typed_dict_catalog.add(type_doc)
+
+    def update(self, typs):
+        for typ in typs:
+            self.add(typ)
+
+    def _get_type_doc_object(self, typ):
+        if isinstance(typ, (EnumDoc, TypedDictDoc)):
+            return typ
+        if isinstance(typ, (TypedDictType, ExtTypedDictType)):
+            return TypedDictDoc.from_TypedDict(typ)
+        if isinstance(typ, EnumType):
+            return EnumDoc.from_Enum(typ)
+        if isinstance(typ, dict):
+            if typ.get('type', None) == 'TypedDict':
+                return TypedDictDoc.from_dict(typ)
+            if typ.get('type', None) == 'Enum':
+                return EnumDoc.from_dict(typ)
+        return None
+
+    def to_dictionary(self):
+        return {
+            'enums': [en.to_dictionary() for en in self.enum_catalog],
+            'typedDicts': [td.to_dictionary() for td in self.typed_dict_catalog]
+        }
+
+
+class TypedDictDoc(Sortable):
+
+    def __init__(self, name='', type='', doc='', items=None):
+        self.name = name
+        self.type = type
+        self.doc = doc
+        self.items = items or []
+
+    @classmethod
+    def from_dict(cls, type_doc):
+        if isinstance(type_doc, dict):
+            return cls(name=type_doc['name'],
+                       type=type_doc['type'],
+                       doc=type_doc['doc'],
+                       items=type_doc['items'])
```

Comment:
Wouldn't `return cls(**type_doc)` work as well? Or does the dict have other keys as well?

Reply:
This is actually a pretty strange alternative constructor for this class because you could just use `TypedDictDoct(**type_doc)` to begin with. Separate `from_TypedDict` makes sense but having that and `from_dict` is rather confusing because both have "dict" in their name but the former handles real types and other apparently operates with already parsed data. I guess the latter, i.e. this method, could be named something like `from_spec` but that still doesn't too useful.